BUG: Raise promotion error if a DType was provided in array coercion

seberg opened the pull request:
This fixes mainly the creation of unstructured void arrays from
scalars, by allowing promotion two unstructured void dtypes.

Structured voids will fail promoting. This promotion error is
now correctly floated out instead of using `object` dtype which
is incorrect when a dtype was passed in.

seberg committed:
BUG: Fix unstructured Void array creation and promotion

This fixes mainly the creation of unstructured void arrays from
scalars, by allowing promotion two unstructured void dtypes.

Structured voids will fail promoting. This promotion error is
now correctly floated out instead of using `object` dtype which
is incorrect when a dtype was passed in.

mattip commented on numpy/core/src/multiarray/array_coercion.c:
```diff
@@ -607,7 +607,7 @@ npy_free_coercion_cache(coercion_cache_obj *next) {
 static NPY_INLINE int
 handle_promotion(PyArray_Descr **out_descr, PyArray_Descr *descr,
-        enum _dtype_discovery_flags *flags)
+        PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags)
```
parameter description in the comment above needs expanding

mattip commented on numpy/core/src/multiarray/array_coercion.c:
```diff
@@ -663,7 +667,7 @@ handle_scalar(
         return -1;
     }
-    if (handle_promotion(out_descr, descr, flags) < 0) {
+    if (handle_promotion(out_descr, descr, fixed_DType, flags) < 0) {
```
Here too it seems the documentation does not describe the parameters correctly, so it is difficult to follow the logic.

mattip commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -258,6 +258,18 @@ string_unicode_common_instance(PyArray_Descr *descr1, PyArray_Descr *descr2)
+    /* Specifically allow unstructured promotion (new in 1.20) */
+    if (descr1->subarray == NULL && descr1->names == NULL &&
+            descr2->subarray == NULL && descr2->names == NULL) {
+        if (descr1->elsize >= descr2->elsize) {
```
Comment?
```suggestion
        /* They must both be flexible descriptor types */
        if (descr1->elsize >= descr2->elsize) {
```

seberg replied:
Added the word flexible to the previous comment, since unstructured voids are what we consider flexible here I guess.

mattip reviewed (commented):
The code itself looks good, but could use some documentation cleanup

seberg committed:
DOC: Fixup promotion/coercion related function comments

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -844,7 +844,29 @@ def test_from_string(self):
+        arr = np.array([b"12345", b"1234"], dtype="V")
+        assert arr.dtype == 'V5'
+        # Check the same for the casting path:
+        arr = np.array([b"1234", b"12345"], dtype="O").astype("V")
```
What does this do? Zero-pad? Undefined-data-pad?

Can you add a test for that?

eric-wieser commented:
I'm not convinced that allowing promotion of void buffers is sensible - the void type is not null-terminated, so `void("\0")` and `void("\0\0")` are different things. If the user wants to null-pad their void data, they can use `np.bytes_`, either directly or by casting to `void` through `bytes..

seberg commented:
Hmmpf, that is a good point. The reason for that change was to make the `np.array` call work because it currently (inexplicitly) does.

I guess I can make the coercion with `np.array([b"a", b"ab"], dtype="V")` just fail and we can probably get away with it. That sounds actually better to me from that perspective. Does that sound fair to you guys?

seberg committed:
BUG: Raise promotion error if a DType was provided in array coercion

This means that we effectively remove certain (strange) array coercions
for unstructured void datatypes.

seberg commented:
OK, I change it to that. We can briefly discuss this, but I think this is the reasonable thing to do. A bare `dtype="V"` should be very strange, and in fact there was a bug here that it went to `object` even though `"V"` was provided and nobody noticed yet, so it doesn't seem likely it affects any important downstream package (and I don't see how they could employ `dtype="V"` reasonably).

eric-wieser commented:
Can we raise an error message that indicates that null padding can be achieved by using bytes instead?

seberg commented:
Not really, since this is a promotion error message you would need special handling just for void somewhere...

eric-wieser commented:
The code right now seems to give the error "invalid type promotion with structured or void datatype(s)". I'm suggesting we tailor that a bit for different types of void dtypes.

Or is your point that the error is swallowed?

seberg committed:
Add error message specific to unstructured void promotion

seberg commented:
Sorry, too much casting :(. For `np.can_cast` I can't raise a custom error, promotion currently indeed raises its own error and that seems fine (we are not likely to write `if can promote_types(...)`).

I added the error, it will also show up for other things (although probably not realistically).

I guess one thing we could think about is creating a `PromotionTypeError` if we ever run into the issue of wanting to be careful about it. In theory our comparison code does run into it. `arr1 == arr2` has (or should have) a `try/except` around promotion and return an array of `False` if it fails.

eric-wieser commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -263,8 +263,17 @@ void_common_instance(PyArray_Descr *descr1, PyArray_Descr *descr2)
+        if (descr1->subarray == NULL && descr1->names == NULL &&
+                descr2->subarray == NULL && descr2->names == NULL) {
+            PyErr_SetString(PyExc_TypeError,
+                    "invalid type promotion with void datatypes of different "
+                    "length. You may use instead `np.bytes_` if the data "
+                    "contains no trailing zero bytes.");
```
Perhaps clearer as:
```suggestion
                    "Invalid type promotion with void datatypes of different "
                    "lengths. Use the `np.bytes_` datatype instead to pad the "
                    "shorter value with trailing zero bytes.");
```

seberg replied:
Yeah, a bit clearer. I guess there are some subtleties when swapping out those types, but I guess if someone already picked unstructured void, they probably know what they are doing, heh...

seberg committed:
Update numpy/core/src/multiarray/dtypemeta.c

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg commented:
Gentle ping, can we put this in?

mattip reviewed (approved):
looks good to me

seberg commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1005,7 +1005,7 @@ def test_promote_types_metadata(self, dtype1, dtype2):
 
     @pytest.mark.parametrize(["dtype1", "dtype2"],
-            [[np.dtype("V6"), np.dtype("V10")],
+            [[np.dtype("V9"), np.dtype("V10")],
```
```suggestion
            [[np.dtype("V6"), np.dtype("V10")],
```

seberg committed:
Update numpy/core/tests/test_numeric.py

seberg commented:
Thanks @mattip, its not a big change and it is holding me up a bit, so I will squash merge it in a bit when the test ran through (undid a leftover change from the first version).
